src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,330 @@
+  setTime(time: number): Promise<number> {
+    return new Promise((resolveOnSetTime, rejectOnSetTime) => {
+      const setTimeCallback: () => Promise<unknown> = () => {
+        return this._videoSourceInfos.then(({ videoSourceBuffer }) => {
+          if (time !== this._thumbnailVideoElement.currentTime) {
+            return this.removeBuffers(videoSourceBuffer, time).then(() => {
+              if (!this._thumbnailTrack) {
+                return rejectOnSetTime(
+                  "VideoThumbnailLoaderError: No thumbnail track given.");
+              }
+
+              const thumbnail: IThumbnailInfo|undefined =
+                arrayFind(this._thumbnailTrack.thumbnailInfos, (t) => {
+                  return t.start <= time && (t.duration + t.start) > time;
+                });
+
+              if (!thumbnail) {
+                return rejectOnSetTime(
+                  "VideoThumbnailLoaderError: Couldn't find thumbnail.");
+              }
+
+              return this.getSegmentData(time, thumbnail.mediaURL).then(({
+                data,
+                fromNetwork,
+              }) => {
+                return videoSourceBuffer
+                  .appendBuffer({
+                    segment: data,
+                    initSegment: this._initSegment,
+                    codec: "null",
+                    timestampOffset: 0,
+                  }).toPromise(PPromise).then(() => {
+                    if (fromNetwork) {
+                      this._bufferedDataCache.push({
+                        start: thumbnail.start,
+                        end: thumbnail.start + thumbnail.duration,
+                        data,
+                      });
+                    }
+                    this._thumbnailVideoElement.currentTime = time;
+                    resolveOnSetTime(this._thumbnailVideoElement.currentTime);
+                    return;
+                  }).catch((err) => {
+                    throw new Error(
+                      "VideoThumbnailLoaderError: Couldn't append buffer :" +
+                      err.message || err
+                    );
+                  });
+              });
+            });
+          }
+          return resolveOnSetTime(this._thumbnailVideoElement.currentTime);
+        });
+      };
+      this._nextCallback = setTimeCallback;
+      /* tslint:disable no-floating-promises */
+      try {
+        this.flush();
+      } catch (e) {
+        return this.dispose().then(() => {
+          throw e;
+        });
+      }
+      /* tslint:enable no-floating-promises */
+    });
+  }
+
+  /**
+   * Update thumbnail track from adaptation
+   * @param {Object} adaptation
+   * @returns {Object}
+   */
+  public updateThumbnailTrack(adaptation: Adaptation): IThumbnailTrack {
+    if (this._onlyTrickMode && !adaptation.trickModeTrack) {
+      throw new Error("VideoThumbnailLoaderError: No provided trick mode track.");
+    }
+
+    const trickModeTrack = adaptation.trickModeTrack || adaptation;
```

Comment:
Just take the trickModeTrack and not the parent Adaptation if you do not use it.

That way, you deport that logic of playing the thumbnail track or the video track to the UI and there's not magic involved.